filament/src/details/MorphTargets.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_DETAILS_MORPHTARGETS_H
+#define TNT_FILAMENT_DETAILS_MORPHTARGETS_H
+
+#include "upcast.h"
+
+#include <filament/MorphTargets.h>
+
+#include "private/backend/DriverApiForward.h"
+
+#include "private/backend/SamplerGroup.h"
+
+#include <backend/Handle.h>
+
+#include <utils/Allocator.h>
+
+#include <vector>
+
+namespace filament {
+
+class FEngine;
+
+class FMorphTargets : public MorphTargets {
+public:
+    FMorphTargets(FEngine& engine, const Builder& builder);
+
+    // frees driver resources, object becomes invalid
+    void terminate(FEngine& engine);
+
+    void setPositionsAt(int targetIndex, math::float3 const* positions, int count);
+    void setPositionsAt(int targetIndex, math::float4 const* positions, int count);
+    void setTangentsAt(int targetIndex, math::short4 const* tangents, int count);
+    inline auto getVertexCount() const noexcept { return mVertexCount; }
+    inline auto getCount() const noexcept { return mCount; }
+
+private:
+    friend class FView;
+    friend class RenderPass;
+
+    void commit(FEngine& engine) const noexcept;
+    void bind(backend::DriverApi& driver) const noexcept;
+
+    class Target {
+    public:
+        explicit Target(int size) noexcept;
+
+        void initialize();
+        void terminate();
+
+        void setPositions(math::float3 const* positions, int count) noexcept;
+        void setPositions(math::float4 const* positions, int count) noexcept;
+        void setTangents(math::short4 const* tangents, int count) noexcept;
+
+        bool isAnyDirty(uint8_t dirtyFlags) const noexcept;
+        void clearDirty() const noexcept;
+
+        inline auto getSize() const noexcept { return mSize; }
+        inline auto getPositions() const noexcept { return mPositions; }
+        inline auto getTangents() const noexcept { return mTangents; }
+
+    private:
+        int mSize;
+        utils::HeapAllocator mAllocator;
```

Comment:
Unless I'm mistaken, we know upfront the number of targets and all Target have the same size, am I correct?
If so, we don't need an allocator per target, all we need is a single malloc() to hold all the Target's data.

Also are the `Target` retained for the whole lifetime of `MorphTarget`? (because it's using quite a bit of memory potentially), are they freed once the data is uploaded into the texture/gpu?

Reply:
Yes you are very correct. But if we don't keep targets, you have to upload target data to buffer every time. I though that it can waste the memory bandwidth. If it's find upload target data to buffer right away, we don't have to target data in staging memory.

You are right. I will figure out to release it.